internal/users/idgenerator.go

```diff
@@ -0,0 +1,400 @@
+			continue
+		}
+
+		if id > maxSuggestedID {
+			log.Warningf(context.Background(),
+				"Generated ID %d is outside the signed 32-bit range, "+
+					"it may not work as expected", id)
+		}
+
+		g.pendingIDs = append(g.pendingIDs, id)
+		cleanup = func() {
+			idx := slices.Index(g.pendingIDs, id)
+			g.pendingIDs = append(g.pendingIDs[:idx], g.pendingIDs[idx+1:]...)
+		}
+		return id, cleanup, nil
+	}
+
+	return 0, nil, fmt.Errorf("failed to find a valid %s for after %d attempts",
+		args.idType, maxAttempts)
+}
+
+// isReservedID checks if the ID is a value is not a linux system reserved value.
+// Note that we are not listing here the system IDs (1…999), as this the job
+// for the [Manager], being a wrong configuration.
+// See: https://systemd.io/UIDS-GIDS/
+func isReservedID(id uint32) bool {
+	switch id {
+	case rootID:
+		// The root super-user.
+		log.Warningf(context.Background(),
+			"ID %d cannot be used: it is the root super-user.", id)
+		return false
+
+	case nobodyID:
+		// Nobody user.
+		log.Warningf(context.Background(),
+			"ID %d cannot be used: it is nobody user.", id)
+		return false
+
+	case uidT32MinusOne:
+		// uid_t-1 (32): Special non-valid ID for `setresuid` and `chown`.
+		log.Warningf(context.Background(),
+			"ID %d cannot be used: it is uid_t-1 (32bit).")
+		return false
+
+	case uidT16MinusOne:
+		// uid_t-1 (16): As before, but comes from legacy 16bit programs.
+		log.Warningf(context.Background(),
+			"ID %d cannot be used: it is uid_t-1 (16bit)", id)
+		return false
+
+	default:
+		return true
+	}
+}
+
+// adjustIDForSafeRanges verifies if the ID value can be safely used, by
+// checking if it's part of any of the well known the ID ranges that can't be
+// used, as per being part of the linux (systemd) reserved ranges.
+// See (again) https://systemd.io/UIDS-GIDS/
+func adjustIDForSafeRanges(id uint32) uint32 {
+	// Human users (homed) (nss-systemd)
+	if id >= nssSystemdHomedMin && id <= nssSystemdHomedMax {
+		log.Warningf(context.Background(),
+			"ID %d is within the nss-systemd range, it cannot be used", id)
+		return systemdContainersUsersMax + 1
+	}
+
+	// Host users mapped into containers (systemd-nspawn) - adjacent to homed!
+	if id >= systemdContainersUsersMin && id <= systemdContainersUsersMax {
+		log.Warningf(context.Background(),
+			"ID %d is within the systemd-nspawn range, it cannot be used", id)
+		return systemdContainersUsersMax + 1
+	}
+
+	// Dynamic service users (nss-systemd)
+	if id >= nssSystemdDynamicServiceUsersMin && id <= nssSystemdDynamicServiceUsersMax {
+		log.Warningf(context.Background(),
+			"ID %d is within the nss-systemd range, it cannot be used", id)
+		return nssSystemdDynamicServiceUsersMax + 1
+	}
+
+	// Container UID ranges (nss-systemd)
+	if id >= nssSystemdContainerMin && id <= nssSystemdContainerMax {
+		log.Warningf(context.Background(),
+			"ID %d is within the nss-systemd range, it cannot be used", id)
```

Comment:
I would print these messages on info or (if you want them to be printed by default) notice level, not warning, because it doesn't indicate a problem.

I would also rephrase it a bit to make clear how this situation is handled (that we're picking another the next available ID):

```suggestion
		log.Noticef(context.Background(),
			"ID %d is within a range used by systemd and cannot be used; skipping to the next available ID after %d",
			id, systemdContainersUsersMax)
```